dwave/optimization/src/nodes/manipulation.cpp

```diff
@@ -18,6 +18,150 @@
 
 namespace dwave::optimization {
 
+std::vector<ssize_t> make_concatenate_shape(std::span<ArrayNode*> array_ptrs, ssize_t axis) {
+    // One or more arrays must be given
+    if (array_ptrs.size() < 1) {
```

Comment:
fwiw `NumPy` handles this case
```python
In [2]: np.concatenate(([0, 1], [2, 3]))
Out[2]: array([0, 1, 2, 3])

In [3]: np.concatenate(([0, 1],))
Out[3]: array([0, 1])
```
so either we should support it at the C++ level, or the Python level. Python support could be
```python
def concatenate(array_likes, axis=0):
    if len(array_likes) == 1:
        return array_likes[0]
```
or C++ support would effectively make `ConcatenateNode` a `CopyNode` (https://github.com/dwavesystems/dwave-optimization/issues/16).

Reply:
Now handled in the python, see [[Add Concatenate to mathematical](https://github.com/dwavesystems/dwave-optimization/pull/153/commits/cb1541999094a1b03aebc457823beaa0c8a14488)](https://github.com/dwavesystems/dwave-optimization/pull/153/commits/cb1541999094a1b03aebc457823beaa0c8a14488)

Reply:
Hmm actually not fixed... numpy seems to cover many different cases

```
>>> a = np.arange(10).reshape((2,5))
>>> np.concatenate(a)
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

>>> np.concatenate((a))
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

>>> np.concatenate((a,))
array([[0, 1, 2, 3, 4],
           [5, 6, 7, 8, 9]])
```

Also works with lists
```
>>> b = np.arange(10).reshape((2,5))
>>> np.concatenate([b])
array([[0, 1, 2, 3, 4],
           [5, 6, 7, 8, 9]])
```

We'll have to resort to some type checking.

```
a = np.array([0,1])
np.concatenate((a))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: zero-dimensional arrays cannot be concatenated
```

Reply:
We don't need to handle every case NumPy does, we just want to be a subset

Reply:
Ah okay, then Ill add the one for single array input

Reply:
The following is now supported and acts like a copy
```
a = model.constant([0,1])
concatenate((a,)) 
```

Since we require ArraySymbols as predecessors, we should either disallow this or construct a `constant` symbol from it. Wdyt? 
```
concatenate([0,1]) 
```